patroni/postgresql/__init__.py

```diff
@@ -462,24 +465,33 @@ def start(self, timeout=None, task=None, block_callbacks=False, role=None):
         else:
             return None
 
-    def checkpoint(self, connect_kwargs=None):
+    @staticmethod
+    def wait_for_checkpoint(check_not_is_in_recovery, connect_kwargs):
+        """Issue PostgreSQL checkpoint."""
+        with get_connection_cursor(**connect_kwargs) as cur:
+            cur.execute("SET statement_timeout = 0")
+            if check_not_is_in_recovery:
+                cur.execute('SELECT pg_catalog.pg_is_in_recovery()')
+                if cur.fetchone()[0]:
+                    return 'is_in_recovery=true'
+            return cur.execute('CHECKPOINT')
+
+    def checkpoint(self, connect_kwargs=None, timeout=None):
         check_not_is_in_recovery = connect_kwargs is not None
         connect_kwargs = connect_kwargs or self.config.local_connect_kwargs
         for p in ['connect_timeout', 'options']:
             connect_kwargs.pop(p, None)
+        if timeout:
+            connect_kwargs['connect_timeout'] = timeout
+            self.stop_retry.deadline = timeout
         try:
-            with get_connection_cursor(**connect_kwargs) as cur:
-                cur.execute("SET statement_timeout = 0")
-                if check_not_is_in_recovery:
-                    cur.execute('SELECT pg_catalog.pg_is_in_recovery()')
-                    if cur.fetchone()[0]:
-                        return 'is_in_recovery=true'
-                return cur.execute('CHECKPOINT')
-        except psycopg2.Error:
+            return self.stop_retry(self.wait_for_checkpoint, check_not_is_in_recovery, connect_kwargs) if timeout \
```

Comment:
While checkpoint is performed postgres continue working normally.
The only thing we need to apply here - `connect_timeout` and if it failed to connect proceed with the shutdown, we don't need to retry the CHECKPOINT.